File: mex/mex/report/party_ledger_report/party_ledger_report.js
// Copyright (c) 2018, Frappe Technologies Pvt. Ltd. and Contributors
// License: GNU General Public License v3. See license.txt

frappe.query_reports["Party Ledger Report"] = {
    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company"),
            reqd: 1,
        },
        {
            fieldname: "party_type",
            label: __("Party Type"),
            fieldtype: "Autocomplete",
            options: ["Customer", "Supplier"],
            reqd: 1,
            on_change: function () {
                frappe.query_report.set_filter_value("party", "");
            },
        },
        {
            fieldname: "party",
            label: __("Party"),
            fieldtype: "MultiSelectList",
            get_data: function (txt) {
                if (!frappe.query_report.filters) return;

                let party_type = frappe.query_report.get_filter_value("party_type");
                if (!party_type) return;

                return frappe.db.get_link_options(party_type, txt);
            },
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            reqd: 1,
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1,
        },
        {
            fieldname: "group_by",
            label: __("Group by"),
            fieldtype: "Select",
            options: [
                "",
                {
                    label: __("Group by Voucher"),
                    value: "Group by Voucher",
                },
                {
                    label: __("Group by Account"),
                    value: "Group by Account",
                },
            ],
            default: "Group by Voucher",
        },
        {
            fieldname: "show_opening_entries",
            label: __("Show Opening Entries"),
            fieldtype: "Check",
        },
        {
            fieldname: "show_cancelled_entries",
            label: __("Show Cancelled Entries"),
            fieldtype: "Check",
        },
        
        {
            fieldname: "show_remarks",
            label: __("Show Remarks"),
            fieldtype: "Check",
        },
        
    ],
};

erpnext.utils.add_dimensions("Party Ledger", 10);

//# sourceURL=party_ledger.js

File: mex/mex/report/party_ledger_report/party_ledger_report.py
# party_ledger_report.py

import frappe
from frappe import _
from frappe.utils import flt, getdate
from erpnext.accounts.report.general_ledger.general_ledger import GeneralLedger

def execute(filters=None):
    # Default filters if none are provided
    filters = filters or {}

    # Fetch party details from filters
    party_type = filters.get('party_type')  # Customer, Supplier, etc.
    party = filters.get('party')  # Party name

    # Fetch the necessary data for the report
    gl_entries = get_party_ledger_data(party_type, party, filters)

    # Format and return the data
    return get_columns(), gl_entries

def get_columns():
    """Define the columns for the Party Ledger report."""
    return [
        _("Date") + ":Date:100",
        _("Voucher Type") + ":Data:150",
        _("Voucher No") + ":Data:150",
        _("Party Name") + ":Data:200",
        _("Account") + ":Data:150",
        _("Debit") + ":Currency:120",
        _("Credit") + ":Currency:120",
        _("Balance") + ":Currency:120",
    ]

def get_party_ledger_data(party_type, party, filters):
    """Fetch the GL entries based on the given party and filters."""
    conditions = [
        "gle.party_type = %s", 
        "gle.party = %s",
        "gle.docstatus = 1"
    ]
    
    # Add filter conditions based on date range and other filters
    if filters.get("from_date"):
        conditions.append("gle.posting_date >= %s")
    if filters.get("to_date"):
        conditions.append("gle.posting_date <= %s")
    
    # Prepare the SQL query
    query = f"""
        SELECT
            gle.posting_date AS date,
            gle.voucher_type,
            gle.voucher_no,
            gle.party,
            gle.account,
            SUM(CASE WHEN gle.debit > 0 THEN gle.debit ELSE 0 END) AS debit,
            SUM(CASE WHEN gle.credit > 0 THEN gle.credit ELSE 0 END) AS credit,
            SUM(CASE WHEN gle.debit > 0 THEN gle.debit ELSE 0 END) - 
            SUM(CASE WHEN gle.credit > 0 THEN gle.credit ELSE 0 END) AS balance
        FROM
            `tabGL Entry` gle
        WHERE
            {conditions}
        GROUP BY
            gle.posting_date, gle.voucher_type, gle.voucher_no, gle.party, gle.account
        ORDER BY
            gle.posting_date ASC
    """

    # Execute the query with the provided filter values
    result = frappe.db.sql(query, tuple([
        party_type,
        party,
        filters.get('from_date'),
        filters.get('to_date')
    ]), as_dict=True)

    # Return the fetched data for report
    return result
